src/databricks/labs/ucx/factories.py

```diff
@@ -0,0 +1,431 @@
+    SqlBackend,
+    StatementExecutionBackend,
+)
+from databricks.sdk import AccountClient, WorkspaceClient
+
+from databricks.labs.ucx.__about__ import __version__
+from databricks.labs.ucx.account import AccountWorkspaces, WorkspaceInfo
+from databricks.labs.ucx.assessment.aws import AWSResources, run_command
+from databricks.labs.ucx.assessment.azure import AzureServicePrincipalCrawler
+from databricks.labs.ucx.aws.access import AWSResourcePermissions
+from databricks.labs.ucx.aws.credentials import CredentialManager, IamRoleMigration
+from databricks.labs.ucx.azure.access import AzureResourcePermissions
+from databricks.labs.ucx.azure.credentials import (
+    ServicePrincipalMigration,
+    StorageCredentialManager,
+)
+from databricks.labs.ucx.azure.locations import ExternalLocationsMigration
+from databricks.labs.ucx.azure.resources import AzureAPIClient, AzureResources
+from databricks.labs.ucx.config import WorkspaceConfig
+from databricks.labs.ucx.hive_metastore import ExternalLocations, Mounts, TablesCrawler
+from databricks.labs.ucx.hive_metastore.catalog_schema import CatalogSchema
+from databricks.labs.ucx.hive_metastore.grants import (
+    AzureACL,
+    GrantsCrawler,
+    PrincipalACL,
+)
+from databricks.labs.ucx.hive_metastore.mapping import TableMapping
+from databricks.labs.ucx.hive_metastore.table_migrate import (
+    MigrationStatusRefresher,
+    TablesMigrator,
+)
+from databricks.labs.ucx.hive_metastore.table_move import TableMove
+from databricks.labs.ucx.hive_metastore.udfs import UdfsCrawler
+from databricks.labs.ucx.installer.workflows import WorkflowsInstallation
+from databricks.labs.ucx.source_code.files import Files
+from databricks.labs.ucx.source_code.languages import Languages
+from databricks.labs.ucx.workspace_access.clusters import ClusterAccess
+from databricks.labs.ucx.workspace_access.groups import GroupManager
+from databricks.labs.ucx.workspace_access.manager import PermissionManager
+
+
+class singleton(property):
+    def __init__(self, func):
+        super().__init__(func)
+        self._qualname = func.__qualname__
+
+    def __get__(self, inst, owner=None):
+        if not isinstance(inst, GlobalContext):
+            raise ValueError("bean must be used in a GlobalContext")
+        __tracebackhide__ = True  # pylint: disable=unused-variable
+        if self._qualname in inst.registry:
+            return inst.registry[self._qualname]
+        created = self.fget(inst)
+        inst.registry[self._qualname] = created
+        return created
+
+    def __set__(self, instance, value):
+        if 'PYTEST_CURRENT_TEST' not in os.environ:
+            raise ValueError("Cannot set bean value outside of testing")
+        instance.registry[self._qualname] = value
+
+    def __str__(self):
+        return f"@bean: {self._qualname}"
+
+
+class GlobalContext:
+    # TODO: make flags only available in CLI contexts
+    flags: dict[str, str]
+
+    def __init__(self):
+        self.registry = {}
```

Comment:
Perhaps this should start with an underscore (`_registry`) to indicate it is a private-ish kind of thing?

Reply:
luckily, we don't need it with `@cached_property` anymore :)